solidity/contracts/bridge/VendingMachine.sol

```diff
@@ -0,0 +1,250 @@
+    uint256 public constant FLOATING_POINT_DIVISOR = 1e18;
+
+    IERC20 public immutable tbtcV1;
+    TBTCToken public immutable tbtcV2;
+
+    /// @notice The fee for unminting TBTC v2 back into TBTC v1. The fee is
+    ///         a portion of the amount being unminted multiplied by
+    ///         `FLOATING_POINT_DIVISOR`. For example, a fee of 1000000000000000
+    ///         means that 0.001 of the value being unminted needs to be paid to
+    ///         the `VendingMachine` as an unminting fee.
+    uint256 public unmintFee;
+    uint256 public newUnmintFee;
+    uint256 public unmintFeeChangeInitiated;
+
+    /// @notice The address of a new vending machine. Set only when the update
+    ///         process is pending. Once the update gets finalized, the new
+    ///         vending machine will become an owner of TBTC v2 token.
+    address public newVendingMachine;
+    uint256 public vendingMachineUpdateInitiated;
+
+    event UnmintFeeUpdateStarted(uint256 newUnmintFee, uint256 timestamp);
+    event UnmintFeeUpdated(uint256 newUnmintFee);
+
+    event VendingMachineUpdateStarted(
+        address newVendingMachine,
+        uint256 timestamp
+    );
+    event VendingMachineUpdated(address newVendingMachine);
+
+    event Minted(address recipient, uint256 amount);
+    event Unminted(address recipient, uint256 amount, uint256 fee);
+
+    modifier onlyAfterGovernanceDelay(uint256 changeInitiatedTimestamp) {
+        GovernanceUtils.onlyAfterGovernanceDelay(
+            changeInitiatedTimestamp,
+            GOVERNANCE_DELAY
+        );
+        _;
+    }
+
+    constructor(
+        IERC20 _tbtcV1,
+        TBTCToken _tbtcV2,
+        uint256 _unmintFee
+    ) {
+        tbtcV1 = _tbtcV1;
+        tbtcV2 = _tbtcV2;
+        unmintFee = _unmintFee;
+    }
+
+    /// @notice Mints TBTC v2 to the caller from TBTC v1 with 1:1 ratio.
+    ///         The caller needs to have at least `amount` of TBTC v1 balance
+    ///         approved for transfer to the `VendingMachine` before calling
+    ///         this function.
+    /// @param amount The amount of TBTC v2 to mint from TBTC v1
+    function mint(uint256 amount) external {
+        _mint(msg.sender, amount);
+    }
+
+    /// @notice Mints TBTC v2 to `from` address from TBTC v1 with 1:1 ratio.
+    ///         `from` address needs to have at least `amount` of TBTC v1
+    ///         balance approved for transfer to the `VendingMachine` before
+    ///         calling this function.
+    /// @dev This function is a shortcut for approve + mint. Only TBTC v1
+    ///      caller is allowed and only TBTC v1 is allowed as a token to
+    ///      transfer.
+    /// @param from TBTC v1 token holder minting TBTC v2 tokens
+    /// @param amount The amount of TBTC v2 to mint from TBTC v1
+    /// @param token TBTC v1 token address
+    function receiveApproval(
+        address from,
+        uint256 amount,
+        address token,
+        bytes calldata
```

Comment:
do we need this `calldata` param?

Reply:
Yes, and this `receiveApproval` signature is an industry-standard though I believe not included in any finalized EIP.